src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.cs

```diff
@@ -875,7 +902,7 @@ public static ReadOnlyCollection<TimeZoneInfo> GetSystemTimeZones()
                         {
                             // sort by BaseUtcOffset first and by DisplayName second - this is similar to the Windows Date/Time control panel
                             int comparison = x.BaseUtcOffset.CompareTo(y.BaseUtcOffset);
-                            return comparison == 0 ? string.CompareOrdinal(x.DisplayName, y.DisplayName) : comparison;
+                            return comparison == 0 ? string.CompareOrdinal(x.Id, y.Id) : comparison;
```

Comment:
> return comparison == 0 ? string.CompareOrdinal(x.Id, y.Id) : comparison; [](http://example.com/codeflow?start=27&length=73)

This is wrong. Usually the list is populated to the UI in most common cases and should be sorted according the display name and not the Id.

Reply:
Sorting by DisplayName will hurt performance considerably because this line may be the first call to lazily load the `DisplayNames`. When ordering by `Id`, this sort operation takes `1ms`, when ordering by `DisplayName`, this sort operation takes `>400ms` (on my machine, for slow android phones it will take even longer).

I am wondering what percentage of users that use `GetSystemTimeZones` need the timezones to be sorted on `DisplayName`, if a majority of users don't need to use DisplayName, I think it would be unfair to make all users who grab `GetSystemTimeZones` for other properties to need to take this significant perf hit. Loading `StandardName` and `DaylightName` together take 1/9th of the time that `DisplayName` takes overall (based on timings within `TryPopulateTimeZoneDisplayNamesFromGlobalizationData`). Would it make sense to have a separate function like `GetSystemTimeZones` + `GetSystemTimeZonesSortedByDisplayName`?

Reply:
This is a breaking change. The behavior is documented.

https://learn.microsoft.com/en-us/dotnet/api/system.timezoneinfo.getsystemtimezones?view=net-7.0#remarks

> The collection returned by this method is sorted by UTC offset and, for time zones that have the same UTC offset, by the display name using the current culture.

Reply:
Would adding another API for users that want to grab all system time zones (regardless of ordering) but care more about performance make sense?

Reply:
Apps usually call GetSystemTimeZones once and not in the startup code paths. I never saw any perf concern with that. Why is the perf concern here? do you have a real app/service suffer from the perf because of that? it should be one call cost and then forget it.

Reply:
And I don't think we need to expose a new API except if we have evidence there is a perf issue blocking apps/services.

Reply:
There was this original issue that this PR aims to remedy https://github.com/xamarin/xamarin-android/issues/8050.

Reply:
I posted a question to the guy who opened the issue there. It is not clear to me why he needs to enumerate all time zones to do time zone conversions.